Make 'rawValue' and 'init(rawValue:)' inlinable for non-resilient enums

jrose-apple opened the pull request:
(~including~ only enums in libraries compiled without -enable-resilience)

Small potential perf win...or rather regain, since we apparently used to do this.

[SR-7094](https://bugs.swift.org/browse/SR-7094) / rdar://problem/38030106

jrose-apple commented:
Still needs tests, but I want to see if this happens to improve any benchmarks. (The JIRA has an ad hoc benchmark that could probably be added to the suite too.)

jrose-apple commented:
@swift-ci Please benchmark

slavapestov commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -146,6 +146,11 @@ static VarDecl *deriveRawRepresentable_raw(TypeChecker &tc,
   getterDecl->setBodySynthesizer(&deriveBodyRawRepresentable_raw);
 
+  if (enumDecl->isExhaustive(nullptr) &&
+      enumDecl->getFormalAccessScope(nullptr, true).isPublic()) {
```
Can you add a comment like `/* respectsVersioned */ true`?

slavapestov reviewed (approved):
Looks good. We can do the same for equatable and hashable as well, however last time I looked the synthesized code for the no-payload case was really inefficient with switching over the tag when it could just bitcast the enum value to an integer and compare for equality. Anyway no need to investigate this in your PR.

jrose-apple commented:
Uh, interesting. Maybe things are getting inlined now where it would be better *not* to.

This change also isn't 100% resilient because *in theory* you could decide `init(rawValue:)` should be able to accept additional values and map them to existing ones, and the inlinable version wouldn't do that. But I don't see a good way to mark the initializer and property as inlinable otherwise, and the hypothetical compatibility checker would catch the change easily, so maybe it's still okay.

eeckstein reviewed (approved):
lgtm.
The benchmark results are ok. +/- 10% is most likely noise.

jrose-apple commented:
Okay, I've thought about this more and:

- `rawValue` seems fine to make inlinable all the time
- but `init(rawValue:)` seems dangerous to make that assumption for resilient libraries

which I think means finding all of the examples in the stdlib where it matters and hand-rolling `init(rawValue:)`.

@slavapestov, thoughts?

jrose-apple commented:
Further discovery: there is exactly one enum in the entire stdlib that would benefit from this: FloatingPointSign (the original cause of the bug). I think I'm going to just implement the synthesized members for that and punt on the original question.

jrose-apple commented:
Did that in #16043. Going to update this PR to *only* add `@inlinable` in non-resilient builds.

jrose-apple committed:
Make 'rawValue' and 'init(rawValue:)' inlinable for non-resilient enums

Small potential perf win...or rather regain, since we apparently used
to do this. It's a bit trickier to say whether we should do the same
for resilient enums, since /in theory/ a later version of the library
might decide to use different raw values.

https://bugs.swift.org/browse/SR-7094

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please smoke test compiler performance
